File: android/src/main/kotlin/jyahann/dgis_map_kit/CameraUtils.kt
package jyahann.dgis_map_kit

import ru.dgis.sdk.coordinates.Bearing
import ru.dgis.sdk.coordinates.GeoPoint
import ru.dgis.sdk.map.CameraPosition
import ru.dgis.sdk.map.Tilt
import ru.dgis.sdk.map.Zoom

class CameraUtils {
    companion object {
        fun getCameraPositionToDart(
            cameraPosition: CameraPosition,
        ): Map<String, Any> {
            return mapOf(
                "position" to getPositionToDart(cameraPosition.point),
                "zoom" to cameraPosition.zoom.value,
                "tilt" to cameraPosition.tilt.value,
                "bearing" to cameraPosition.bearing.value,
            )
        }

        fun getPositionToDart(geoPoint: GeoPoint) : Map<String, Any> {
            return mapOf(
                "lat" to geoPoint.latitude.value,
                "long" to geoPoint.longitude.value,
            )
        }

        fun getCameraPositionFromDart(
            cameraPosition: Map<String, Any>,
        ): CameraPosition {
            return CameraPosition(
                point = getGeoPointFromDart(cameraPosition["position"] as Map<String, Any>),
                zoom = Zoom((cameraPosition["zoom"] as Double).toFloat()),
                tilt = Tilt((cameraPosition["tilt"] as Double).toFloat()),
                bearing = Bearing(cameraPosition["bearing"] as Double),
            )
        }

        fun getGeoPointFromDart(position: Map<String, Any>): GeoPoint {
            return GeoPoint(
                latitude = position["lat"] as Double,
                longitude = position["long"] as Double
            )
        }
    }
}

File: ios/Classes/CameraUtils.swift
import DGis;

class CameraUtils {
    static func getCameraPositionToDart(
        cameraPosition position: CameraPosition
    ) -> Dictionary<String, Any> {
        return [
            "position": getPositionToDart(
                geoPoint: position.point
            ),
            "zoom": position.zoom.value,
            "tilt": position.tilt.value,
            "bearing": position.bearing.value
        ];
    }
    
    static func getPositionToDart(
        geoPoint point: GeoPoint
    ) -> Dictionary<String, Any> {
        return [
            "lat": point.latitude.value,
            "long": point.longitude.value
        ];
    }
    
    static func getCameraPositionFromDart(
        cameraPosition position: Dictionary<String, Any>
    ) -> CameraPosition {
        return CameraPosition(
            point: getGeoPointFromDart(
                coordinates: position["position"] as! Dictionary<String, Any>
            ),
            zoom: Zoom(
                floatLiteral: Float(position["zoom"] as! Double)
            ),
            tilt: Tilt(
                floatLiteral: Float(position["tilt"] as! Double)
            ),
            bearing: Bearing(
                floatLiteral: position["bearing"] as! Double
            )
        );
    }
    
    static func getGeoPointFromDart(
        coordinates coords: Dictionary<String, Any>
    ) -> GeoPoint {
        return GeoPoint(
            latitude: coords["lat"] as! Double,
            longitude: coords["long"] as! Double
        );
    }
}
